From: Jeff Layton <jlayton@redhat.com>
Date: Mon, 18 Aug 2008 08:44:28 -0400
Subject: [nfs] v4: credential ref leak in nfs4_get_state_owner
Message-id: 1219063469-27779-3-git-send-email-jlayton@redhat.com
O-Subject: [RHEL5.3 PATCH 2/3] BZ#441884: NFSv4: Fix a credential reference leak in nfs4_get_state_owner()
Bugzilla: 441884
RH-Acked-by: Peter Staubach <staubach@redhat.com>

From: Trond Myklebust <Trond.Myklebust@netapp.com>

Upstream commit 27b3f949b769a208e2849d28e7ad64cadac5d0e3

diff --git a/fs/nfs/nfs4state.c b/fs/nfs/nfs4state.c
index babe598..4349907 100644
--- a/fs/nfs/nfs4state.c
+++ b/fs/nfs/nfs4state.c
@@ -83,7 +83,7 @@ nfs4_client_grab_unused(struct nfs_client *clp, struct rpc_cred *cred)
 	if (!list_empty(&clp->cl_unused)) {
 		sp = list_entry(clp->cl_unused.next, struct nfs4_state_owner, so_list);
 		atomic_inc(&sp->so_count);
-		sp->so_cred = cred;
+		sp->so_cred = get_rpccred(cred);
 		list_move(&sp->so_list, &clp->cl_state_owners);
 		clp->cl_nunused--;
 	}
@@ -175,7 +175,6 @@ struct nfs4_state_owner *nfs4_get_state_owner(struct nfs_server *server, struct
 	struct nfs_client *clp = server->nfs_client;
 	struct nfs4_state_owner *sp, *new;
 
-	get_rpccred(cred);
 	new = nfs4_alloc_state_owner();
 	spin_lock(&clp->cl_lock);
 	sp = nfs4_find_state_owner(clp, cred);
@@ -185,7 +184,7 @@ struct nfs4_state_owner *nfs4_get_state_owner(struct nfs_server *server, struct
 		list_add(&new->so_list, &clp->cl_state_owners);
 		new->so_client = clp;
 		new->so_id = nfs4_alloc_lockowner_id(clp);
-		new->so_cred = cred;
+		new->so_cred = get_rpccred(cred);
 		sp = new;
 		new = NULL;
 	}
@@ -193,7 +192,6 @@ struct nfs4_state_owner *nfs4_get_state_owner(struct nfs_server *server, struct
 	kfree(new);
 	if (sp != NULL)
 		return sp;
-	put_rpccred(cred);
 	return NULL;
 }
 
